### PR TITLE
Obsoleting or removing methods that can potentially block indefinitely

### PR DESCRIPTION
**Pull request by mhowlett**
@treziac @edenhill @ewencp 

People who use variants of these methods with an infinite timeout usually don't realize they can't cancel out of them or even that they generally might want to - better if they don't exist.

Actually we should add variants that take a `CancellationToken` as well, but that can happen in a latter PR.

### CODE REVIEW

File: src/Confluent.Kafka/Consumer.cs [hunk 1]
```diff
@@ -151,21 +151,14 @@ public bool Consume(out Message<TKey, TValue> message, int millisecondsTimeout)
...
-            => Consume(out message, -1);
 
         /// <summary>
         ///     Poll for new consumer events, including new messages
         ///     ready to be consumed (which will trigger the OnMessage
         ///     event).
         /// </summary>
         /// <param name="millisecondsTimeout"> 
-        ///     The maximum time to block (in milliseconds).
+        ///     The maximum time to block (in milliseconds), or -1 to block indefinitely.
```
Comment (treziac) [inline src/Confluent.Kafka/Consumer.cs]: Add that -1 should not typically be used. Same remarks for the Poll(TimeSpan) : TimeSpan.Infinite should typically not be used

File: src/Confluent.Kafka/Consumer.cs [hunk 2]
```diff
@@ -195,12 +188,13 @@ public void Poll(TimeSpan timeout)
...
         /// <summary>
         ///     Poll for new consumer events, including new messages
-        ///     ready to be consumed(which will trigger the OnMessage
+        ///     ready to be consumed (which will trigger the OnMessage
         ///     event).
         /// </summary> 
         /// <remarks>
         ///     Blocks indefinitely until a new event is ready.
         /// </remarks>
+        [Obsolete("Indefinite polling cannot be cancelled so it's usually not what you want. Removing this method to discourage.", false)]
```
Comment (treziac) [inline src/Confluent.Kafka/Consumer.cs]: Just "Use Poll overload with finite timeout" is enough I think (obsolete is already discouraging)

File: src/Confluent.Kafka/Consumer.cs [hunk 3]
```diff
@@ -1015,7 +995,7 @@ public void Poll(TimeSpan timeout)
         ///     event).
         /// </summary>
         /// <param name="millisecondsTimeout"> 
-        ///     The maximum time to block (in milliseconds)
+        ///     The maximum time to block (in milliseconds), or -1 to block indefinitely.
```
Comment (treziac) [inline src/Confluent.Kafka/Consumer.cs]: Same remark

File: src/Confluent.Kafka/Consumer.cs [hunk 4]
```diff
@@ -1034,6 +1014,7 @@ public void Poll(int millisecondsTimeout)
         /// <remarks>
         ///     Blocks indefinitely until a new event is ready.
         /// </remarks>
+        [Obsolete("Indefinite polling cannot be cancelled so it's usually not what you want. Removing this method to discourage.", false)]
         public void Poll()
```
Comment (treziac) [inline src/Confluent.Kafka/Consumer.cs]: Same remark

File: src/Confluent.Kafka/Producer.cs [hunk 5]
```diff
@@ -360,6 +360,7 @@ public int Poll(TimeSpan timeout)
         /// <returns>
         ///     Returns the number of events served.
         /// </returns>
+        [Obsolete("Indefinite polling cannot be cancelled so it's usually not what you want. Removing this method to discourage.", false)]
```
Comment (treziac) [inline src/Confluent.Kafka/Producer.cs]: Same remark

File: src/Confluent.Kafka/Producer.cs [hunk 6]
```diff
@@ -624,6 +643,7 @@ public int Flush(int millisecondsTimeout)
         /// <returns>
         ///     Refer to <see cref="Flush(int)" />.
         /// </returns>
+        [Obsolete("This method cannot be cancelled so it's usually not what you want. Removing this method to discourage.", false)]
```
Comment (treziac) [inline src/Confluent.Kafka/Producer.cs]: Same remark

File: src/Confluent.Kafka/Producer.cs [hunk 7]
```diff
@@ -616,6 +617,24 @@ public int Flush(int millisecondsTimeout)
...
         /// <summary>
+        ///     Wait until all outstanding produce requests and delievery report
+        ///     callbacks are completed. Refer to <see cref="Flush(int)" /> for
+        ///     more information.
+        /// 
+        ///     [UNSTABLE-API] - the semantics and/or type of the return value is
+        ///     subject to change.
+        /// </summary>
+        /// <param name="timeout">
+        ///     The maximum length of time to block.
```
Comment (treziac) [inline src/Confluent.Kafka/Producer.cs]: Same remark (avoid TimeStamp.Timeout which may block kafka thread or something like that)
The maximum time to block (no length of)

File: src/Confluent.Kafka/Producer.cs [hunk 8]
```diff
@@ -1211,20 +1223,29 @@ public int Flush(int millisecondsTimeout)
...
+        ///     callbacks are completed. Refer to <see cref="Flush(int)" /> for
+        ///     more information.
+        /// 
         ///     [UNSTABLE-API] - the semantics and/or type of the return value is
         ///     subject to change.
         /// </summary>
-        public int Flush()
-            => producer.Flush();
+        /// <param name="timeout">
+        ///     The maximum length of time to block.
```
Comment (treziac) [inline src/Confluent.Kafka/Producer.cs]: Same remark

File: src/Confluent.Kafka/Consumer.cs (hunk 2 above)
Comment (mhowlett) [inline src/Confluent.Kafka/Consumer.cs]: i'd agree with that

**Pull request merged**